app.go

```diff
@@ -113,6 +113,40 @@ func (o provideOption) String() string {
 	return fmt.Sprintf("fx.Provide(%s)", strings.Join(items, ", "))
 }
 
+// Supply makes the given values available for dependency resolution as if
```

Comment:
The logic seems good to me, though i'm wondering if extending Provides would have been a better user experience. Would something like ValueProvide be more fitting than Supply?

Reply:
Why can't Provide() take both functions and/or values?

Reply:
On the two points raised:

- There is no precedent so far about overloading the "provider" term. Since the connotation associated with a "provider" is that it's a function, a ___Provider that doesn't consume a function would be surprising.
- `Provide()` shouldn't be modified to implicitly provide values, as it would not be able to differentiate between (a) regular provider functions; and (b) function values that are to be provided. For example: given `type MyFunc func(context.Context) error`, one might choose to `fx.Provide(func() MyFunc { return ... })`, and thus use `MyFunc` from other providers or invokes.

(side note: that last part is a great alternative to traditional service-repository patterns)

Reply:
Historical context: `Provide` used to accept both, functions and
values. If the argument was not a function, it would stick it into the
container as-is using the reflected type.

An issue this caused was that the "most specific" type, as determined
by `reflect.TypeOf` would be used. This meant `fx.Provide(new(bytes.Buffer))`
would be used as a `*bytes.Buffer` instead of `io.Writer`. The obvious
fix for that was `fx.Provide(func() io.Writer { .. })`. We decided that
debugging issues from that design before you got to the `func()` based
solution would be frustrating, so we opted for `func()`-only with
explicit types everywhere, with the intention of adding value-based
provides if there was enough need for it.

Now that we've had a fair amount of production experience with Fx, I'm
supportive of this. A separate entrypoint makes sense, and fx.Supply
seems like a good name for it.

To @roberth-k's second point: We definitely have internal Fx modules
that produce naked `func`s into the container as well as newtype
wrapped funcs.

Reply:
> would be used as a `*bytes.Buffer` instead of `io.Writer`

@abhinav slightly off topic, but why can't Fx do both? When something is expecting  an interface as input, it's easy to scan all objects in the DI context and check if they can be duck typed to that interface. The current behavior of most specific type always seemed counterintuitive to me (coming from DIs in Java, like Spring, which happily matches on interfaces).

Reply:
@yurishkuro we chose to make this explicit specifically because Go
interfaces are implicit: types do not declare which interfaces they
implement. This would have the effect of matching types to interfaces
users did not intend (possibly non-deterministically). Given that Fx's
injection all happens at runtime, debugging issues arising from that
would be more difficult. There was some prior discussion on this in https://github.com/uber-go/fx/issues/608#issuecomment-352881873.

If we can come up with a safe method of opting into interface
conversion on a case-by-case basis, that would be more desirable. #673
gets close here—we can finish implementing that. There was a prior
attempt at implementing it but the behavior wasn't quite right so the
change was rolled back.